src/shared/pkcs11-util.c

```diff
@@ -523,6 +523,272 @@ int pkcs11_token_find_x509_certificate(
+                return 0;
+        }
+
+        switch (key_type) {
+        case CKK_RSA: {
+                CK_ATTRIBUTE rsa_attributes[] = {
+                        { CKA_MODULUS,         NULL_PTR, 0 },
+                        { CKA_PUBLIC_EXPONENT, NULL_PTR, 0 }
+                };
+                _cleanup_free_ void *modulus = NULL, *public_exponent = NULL;
+
+                rv = m->C_GetAttributeValue(session, object, rsa_attributes, ELEMENTSOF(rsa_attributes));
+                if (rv != CKR_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                "Failed to get size of attributes of an RSA public key: %s", sym_p11_kit_strerror(rv));
+
+                if (rsa_attributes[0].ulValueLen != 0) {
+                        modulus = malloc(rsa_attributes[0].ulValueLen);
+                        if (!modulus)
+                                return log_oom();
+
+                        rsa_attributes[0].pValue = modulus;
+                } else
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "An RSA public key has empty CKA_MODULUS.");
+
+                if (rsa_attributes[1].ulValueLen != 0) {
+                        public_exponent = malloc(rsa_attributes[1].ulValueLen);
+                        if (!public_exponent)
+                                return log_oom();
+
+                        rsa_attributes[1].pValue = public_exponent;
+                } else
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "An RSA public key has empty CKA_PUBLIC_EXPONENT.");
+
+                rv = m->C_GetAttributeValue(session, object, rsa_attributes, ELEMENTSOF(rsa_attributes));
+                if (rv != CKR_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                "Failed to get attributes of an RSA public key: %s", sym_p11_kit_strerror(rv));
+
+                size_t n_size = rsa_attributes[0].ulValueLen, e_size = rsa_attributes[1].ulValueLen;
+                r = rsa_pkey_from_n_e(rsa_attributes[0].pValue, n_size, rsa_attributes[1].pValue, e_size, &pkey);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create an EVP_PKEY from RSA parameters.");
+
+                break;
+        }
+        case CKK_EC: {
+                CK_ATTRIBUTE ec_attributes[] = {
+                        { CKA_EC_PARAMS, NULL_PTR, 0 },
+                        { CKA_EC_POINT,  NULL_PTR, 0 }
+                };
+                _cleanup_free_ void *ec_group = NULL, *ec_point = NULL;
+
+                rv = m->C_GetAttributeValue(session, object, ec_attributes, ELEMENTSOF(ec_attributes));
+                if (rv != CKR_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                "Failed to get size of attributes of an EC public key: %s", sym_p11_kit_strerror(rv));
+
+                if (ec_attributes[0].ulValueLen != 0) {
```

Comment:
this fields a bit weird. just invert the if check, so that the exceptional code path (i.e. the error) remains indented. And the do the "main" code path (i.e. the malloc()) in the original indentation level